### PR TITLE
 ACTIONX container

### PR DESCRIPTION
**Pull request by joakim-hove**
This PR implements a container to hold/manage ACTIONX keywords. There is much work remaining before we have a working ACTIONX implementation, but this should be ready to be merged now.

What is missing includes

1. A parser to parse the ACTIONX expressions.
2. Generate the context for the evaluation.
3. Apply the keywords in the action.

I'll proceed one PR at a time.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Review?

File: opm/parser/eclipse/EclipseState/Schedule/Actions.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <string>
+#include <ctime>
+#include <map>
+
+#include <opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp>
+
+namespace Opm {
+
+class ActionX;
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/Actions.hpp]: Yes - I'll remove that: the forward was from before I needed the full object and added the `#include`

File: src/opm/parser/eclipse/EclipseState/Schedule/ActionX.cpp [hunk 2]
```diff
@@ -23,27 +23,65 @@
...
-const std::string& ActionX::name() const {
-    return this->m_name;
+
+
+bool ActionX::eval(std::time_t sim_time, const ActionContext& context) {
+    if (!this->ready(sim_time))
+        return false;
+    bool result = true;
+
+    if (result) {
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/ActionX.cpp]: Yes - `result` will in the future be `false`. The semantis of the `ACTIONX` keyword is essentially:
```
if (should_this_be_run()) {
   if (action_eval()) {
      apply_keywords;
   }
}
```
where currently *only* the `should_this_be_run()` part is implemented, and `action_eval()` is hadcoded to `true`.

**Comment by joakim-hove** [comment]
> Very strictly speaking, [...]

OK - I will not argue with the facts here; but going through the code to use `std::size_t` is a bit over my level of pedantry.

File: opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp [hunk 3]
```diff
@@ -49,21 +53,27 @@ namespace Opm {
...
 
-    const std::string& name() const;
-private:
-    std::string m_name;
-    size_t max_run;
-    double max_wait;
+    const std::string name;
+    const size_t max_run;
+    const double min_wait;
+    const std::time_t start_time;
```
Comment (bska) [inline opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp]: I didn't pick up on this before (sorry), but those constant data members may come back to hurt us.  They prevent the compiler from generating assignment operators (i.e., `operator=(const& ActionX& rhs)` and `operator=(ActionX&& rhs)`) which, in turn, will limit the available operations on containers of `ActionX` (e.g., class `Opm::Actions`).

If you need these members to behave as `const` I'd probably recommend making them `private` and then providing `const` qualified accessor functions for them instead.

Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp]: Ahhhh - the boilerplate - you got to love it; this language has a certain feeling of quiltwork by commitee ..,

> They prevent the compiler from generating assignment operators (i.e., operator=(const& ActionX& rhs) and operator=(ActionX&& rhs)) 

I will not argue this - your knowledge of this far outweighs mine; but I do not understand it. The `operator= (const& ActionX& rhs)` should be perfectly safe to generate? I don't know the semantics of "moving out of"  a value type, but that should in my opinion be perfectly reducable to a no-op.

Anyway - I take your word at face value and will update with private members and so on.

Comment (bska) [reply opm/parser/eclipse/EclipseState/Schedule/ActionX.hpp]: > > They prevent the compiler from generating assignment operators (i.e., operator=(const& ActionX& rhs) and operator=(ActionX&& rhs))

Doh, that should obviously be
```C++
operator=(const ActionX& rhs)
```
Sorry about the extra ampersand.

> I do not understand it. The `operator= (const ActionX& rhs)` should be perfectly safe to generate?

The default/compiler-generated assignment operator behaves like the following pseudo code
```C++
ActionX& ActionX::operator=(const ActionX& rhs)
{
    for (auto& member : this->members()) {
        this->member = rhs.member;
    }

    return *this;
}
```
If any of the members is `const`, then it cannot be assigned which means that the code above cannot do what it's supposed to do and is therefore not generated at all.  The default move-assignment operator looks the same except that the loop body is
```
this->member = std::move(rhs.member);
```
This matters if one of the members is a vector of databases, but falls back to simple assignment/copy if the member is just a built-in arithmetic type like an `int`.

**Comment by joakim-hove** [comment]
> [...] and will update with private members and so on.

Done

**Comment by joakim-hove** [comment]
jenkins build this please

**Pull request merged**